Add oom test

jbartosik opened the pull request.

krzysied commented on vertical-pod-autoscaler/e2e/v1/full_vpa.go:
```diff
@@ -120,6 +120,52 @@ var _ = FullVpaE2eDescribe("Pods under VPA", func() {
+		vpaClientSet *vpa_clientset.Clientset
+		vpaCRD       *vpa_types.VerticalPodAutoscaler
+	)
+	replicas := 3
```
const?

jbartosik replied:
Done

krzysied commented on vertical-pod-autoscaler/e2e/v1/autoscaling_utils.go:
```diff
@@ -427,3 +429,25 @@ func runServiceAndWorkloadForResourceConsumer(c clientset.Interface, ns, name st
+	dpConfig := testutils.DeploymentConfig{
+		RCConfig: rcConfig,
+	}
+	ginkgo.By(fmt.Sprintf("creating deployment %s in namespace %s", dpConfig.Name, dpConfig.Namespace))
```
nit: Capital letter.

jbartosik replied:
Done.

bskiba commented:
Just a side note: Please remember to add to v1beta2 as well as long as we have it.

jbartosik commented on vertical-pod-autoscaler/e2e/v1/full_vpa.go:
```diff
@@ -120,6 +120,52 @@ var _ = FullVpaE2eDescribe("Pods under VPA", func() {
+	)
+	replicas := 3
+
+	// This schedules AfterEach block that needs to run after the AfterEach above and
```
Left over from c&p; removed

jbartosik commented on vertical-pod-autoscaler/e2e/v1/full_vpa.go:
```diff
@@ -120,6 +120,52 @@ var _ = FullVpaE2eDescribe("Pods under VPA", func() {
+			ParseQuantityOrDie("1400Mi"), ParseQuantityOrDie("10000Mi"))
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+	})
+
```
Removed

jbartosik commented on vertical-pod-autoscaler/e2e/v1/full_vpa.go:
```diff
@@ -120,6 +120,52 @@ var _ = FullVpaE2eDescribe("Pods under VPA", func() {
+		vpaClient := vpaClientSet.AutoscalingV1()
+		_, err = vpaClient.VerticalPodAutoscalers(ns).Create(vpaCRD)
+		gomega.Expect(err).NotTo(gomega.HaveOccurred())
+
```
Removed

jbartosik commented on vertical-pod-autoscaler/e2e/utils/ooming_resource_consumer/Dockerfile:
```diff
@@ -0,0 +1,2 @@
+FROM gcr.io/google-containers/stress:v1
+ENTRYPOINT ["/stress", "--mem-total", "10000000000", "-logtostderr"]
```
I changed to `--logtostderr` and it still works so I think it doesn't matter. I'll keep version with double dash to make it prettier

jbartosik commented on vertical-pod-autoscaler/e2e/v1/autoscaling_utils.go:
```diff
@@ -427,3 +429,25 @@ func runServiceAndWorkloadForResourceConsumer(c clientset.Interface, ns, name st
 }
+
+func runOomingReplicationController(c clientset.Interface, ns, name string, replicas int) {
+	ginkgo.By(fmt.Sprintf("Running fixed resource consuming RC %s with %v replicas", name, replicas))
```
Done

jbartosik commented on vertical-pod-autoscaler/e2e/v1/full_vpa.go:
```diff
@@ -120,6 +120,52 @@ var _ = FullVpaE2eDescribe("Pods under VPA", func() {
 
+var _ = FullVpaE2eDescribe("OOMing pods under VPA", func() {
+	var (
+		vpaClientSet *vpa_clientset.Clientset
```
We don't need them. But if we want to have more tests on OOMing pods it will be easier

jbartosik commented:
I forgot that Github sends my replies to comments immediately rather than batching them and waiting for me to send the batch; I now pushed fixed commits; I'll push v1beta2 test later today

jbartosik commented:
@bskiba I'm surprised - when I try to run the same test using `v1beta2` API it consistently fails (times out).

bskiba commented:
Weird, that should behave exactly the same...

jbartosik commented:
@krzysied please take a look.

Tests I'm adding pass now. They were failing because after OOMing pods sometimes stay around for longer time. The test wasn't filtering out failed pods and the old pods had wrong requests. I added filtering and now test consistently passes.

I also changed the loop to iterate over all pods when comparing requests rather than end after finding first pod that doesn't match. This change helped me a lot when I was debugging and I think it's worth keeping

jbartosik commented:
@bskiba , @krzysied please take a look. I added another commit to the PR - it changes timeout on the OOM test so that the test still passes. But the test modified to use an image that doesn't OOM (but uses enough memory to trigger growth) fails

krzysied commented on vertical-pod-autoscaler/e2e/v1/full_vpa.go:
```diff
@@ -135,18 +179,19 @@ func waitForPodsMatch(f *framework.Framework, listOptions metav1.ListOptions, ma
 		for _, pod := range podList.Items {
 			if !matcher(pod) {
-				return false, nil
+				result = false
```
Why the return command was removed?

jbartosik replied:
I moved it to after the loop. This is to evaluate all the pods (instead of exiting after the first pod that doesn't match). It makes debugging easier (for example test was failing because it didn't filter out failed pods. Which was hard to notice when we logged only about the first pod on the list)

krzysied replied:
Please add comment why we want to have it this way.
TBH, it looks weird without a backstory.

jbartosik replied:
Done

krzysied commented:
/lgtm
/approve
/hold to give @bskiba time for the review

krzysied commented:
no comments
/hold cancel
